File: extensions/vscode/src/commands.ts
// Every command the extension contributes.
//
// The ids and their titles are a public contract — package.json declares them,
// keybindings and the palette use them, and the pane's title bar wires them to
// its icons — so this file registers exactly the list in `contributes.commands`
// plus depwatch.reveal, which the webview calls and nobody types.
//
// Handlers take what they need through `CommandDeps` rather than closing over
// activate()'s locals: the two that change under them (the configuration and
// the scanner, both rebuilt when settings move) come in as getters, so a
// command always runs against the current one.

import * as vscode from 'vscode'
import { relative } from 'node:path'
import { basename } from '../../../src/manifest.js'
import { quadrantSVG } from '../../../src/quadrant.js'
import { gateFailures } from '../../../src/gates.js'
import { INSTALL_GIT, isMissingGit, trend } from '../../../src/trend.js'
import { acceptableCount, type WorkspaceBaseline } from './baseline-file.js'
import type { Config } from './config.js'
import { findManifests, isScannable, type Scanner } from './engine.js'
import { locateDeps } from './locate.js'
import { type ReportPanel, showTrend } from './panel.js'
import type { Results } from './state.js'
import { LENS_BLURB, LENS_LABEL, LENSES } from './totals.js'
import type { FindingsTree } from './tree.js'

export interface CommandDeps {
  log: { show(): void }
  results: Results
  tree: FindingsTree
  panel: ReportPanel
  baseline: WorkspaceBaseline
  secrets: vscode.SecretStorage
  /** The current configuration; rebuilt whenever settings change. */
  cfg: () => Config
  /** The current scanner; rebuilt whenever settings change. */
  scanner: () => Scanner
  scanOne: (path: string, opts?: { deep?: boolean; force?: boolean }) => Promise<void>
  scanAll: (opts?: { deep?: boolean; force?: boolean; quiet?: boolean }) => Promise<void>
  /** Abort the running scan; false when there was none. */
  cancel: () => boolean
}

export function registerCommands(context: vscode.ExtensionContext, deps: CommandDeps): void {
  const { log, results, tree, panel, baseline, cfg, scanner, scanOne, scanAll } = deps

  const register = (name: string, run: (...args: any[]) => unknown) =>
    context.subscriptions.push(vscode.commands.registerCommand(name, run))

  async function activeManifest(): Promise<string | undefined> {
    const active = vscode.window.activeTextEditor?.document
    if (active?.uri.scheme === 'file' && isScannable(active.uri.fsPath, cfg())) return active.uri.fsPath

    const candidates = results.size > 0 ? results.all().map((s) => s.path) : await findManifests(cfg())
    if (candidates.length === 0) {
      vscode.window.showWarningMessage('depwatch: no dependency manifest found.')
      return undefined
    }
    if (candidates.length === 1) return candidates[0]
    return vscode.window
      .showQuickPick(
        candidates.map((path) => ({ label: vscode.workspace.asRelativePath(path), path })),
        { title: 'Which manifest?' },
      )
      .then((pick) => pick?.path)
  }

  // --- scanning ---

  register('depwatch.scanWorkspace', () => scanAll())
  register('depwatch.deepScan', () =>
    vscode.window.withProgress(
      { location: vscode.ProgressLocation.Notification, title: 'depwatch: deep scan' },
      () => scanAll({ deep: true, force: true }),
    ),
  )
  register('depwatch.scanFile', async () => {
    const path = await activeManifest()
    if (path) await scanOne(path, { force: true })
  })

  register('depwatch.cancel', () => {
    if (!deps.cancel()) vscode.window.showInformationMessage('depwatch: no scan is running.')
  })

  register('depwatch.clearCache', async () => {
    await scanner().clear()
    results.clear()
    vscode.window.showInformationMessage('depwatch: registry cache cleared. The next scan will refetch.')
  })

  // --- the pane ---

  register('depwatch.setScopeFile', () => tree.setScope('file'))
  register('depwatch.setScopeProject', () => tree.setScope('project'))
  register('depwatch.expandAll', () => tree.expandAll())
  register('depwatch.showLog', () => log.show())
  register('depwatch.clearFilter', () => tree.setFilter(null))

  // A multi-select quick pick rather than a row of toggle buttons: five
  // quadrants would be five more icons in a title bar that already has enough,
  // and the counts belong next to the choice.
  register('depwatch.filterFindings', async () => {
    const counts = tree.scopeCounts()
    const active = tree.getFilter()
    const items = LENSES.map((lens) => ({
      label: LENS_LABEL[lens],
      description: `${counts[lens]} ${counts[lens] === 1 ? 'dep' : 'deps'}`,
      detail: LENS_BLURB[lens],
      lens,
      picked: active ? active.has(lens) : true,
    }))
    const picked = await vscode.window.showQuickPick(items, {
      canPickMany: true,
      title: 'Show findings for',
      placeHolder: 'Pick the quadrants to show — none picked shows everything',
    })
    if (!picked) return // cancelled: leave the filter alone
    tree.setFilter(picked.length === 0 ? null : new Set(picked.map((p) => p.lens)))
  })

  register('depwatch.reveal', async (path: string, dep: string) => {
    const doc = await vscode.workspace.openTextDocument(vscode.Uri.file(path))
    const editor = await vscode.window.showTextDocument(doc, { preview: true })
    const span = locateDeps(doc.getText(), path, [dep]).get(dep)
    if (!span) return
    const range = new vscode.Range(doc.positionAt(span.start), doc.positionAt(span.end))
    editor.selection = new vscode.Selection(range.start, range.end)
    editor.revealRange(range, vscode.TextEditorRevealType.InCenterIfOutsideViewport)
  })

  // --- the report ---

  register('depwatch.showReport', async () => {
    if (results.size === 0) await scanAll({ quiet: true })
    panel.show()
  })

  register('depwatch.checkGates', async () => {
    if (results.size === 0) await scanAll({ quiet: true })
    if (!cfg().gatesConfigured) {
      const pick = await vscode.window.showInformationMessage(
        'depwatch: no gates configured. Set depwatch.gates.maxLibyears or depwatch.gates.maxReplace to fail on a budget, the same way depwatch check --ci does.',
        'Open settings',
      )
      if (pick) await vscode.commands.executeCommand('workbench.action.openSettings', 'depwatch.gates')
      return
    }
    const failures = results
      .all()
      .flatMap((s) => gateFailures(s.report, cfg().gates).map((f) => `${s.label}: ${f.message}`))
    if (failures.length === 0) {
      vscode.window.showInformationMessage('depwatch: gates pass.')
      return
    }
    const pick = await vscode.window.showWarningMessage(
      `depwatch: ${failures.length} gate(s) failing — ${failures[0]}`,
      'Show report',
    )
    if (pick) panel.show()
  })

  register('depwatch.showTrend', async () => {
    const path = await activeManifest()
    if (!path) return
    const folder = vscode.workspace.getWorkspaceFolder(vscode.Uri.file(path))
    if (!folder) {
      vscode.window.showWarningMessage('depwatch: trend needs the manifest to sit inside a workspace folder.')
      return
    }
    // git wants a repo-relative path, with forward slashes on every platform.
    const rel = relative(folder.uri.fsPath, path).split(/[\\/]/).join('/')
    await vscode.window.withProgress(
      { location: vscode.ProgressLocation.Notification, title: `depwatch: reading history of ${rel}` },
      () => showHistory(rel, folder.uri.fsPath, cfg()),
    )
  })

  // --- exporting ---

  register('depwatch.exportReport', async () => {
    if (results.size === 0) await scanAll({ quiet: true })
    const uri = await vscode.window.showSaveDialog({
      filters: { HTML: ['html'] },
      saveLabel: 'Export report',
      defaultUri: defaultExportUri('depwatch-report.html'),
    })
    if (!uri) return
    await vscode.workspace.fs.writeFile(uri, new TextEncoder().encode(panel.export()))
    const pick = await vscode.window.showInformationMessage(`depwatch: wrote ${basename(uri.fsPath)}`, 'Open')
    if (pick) await vscode.env.openExternal(uri)
  })

  register('depwatch.exportChart', async () => {
    const path = await activeManifest()
    if (!path) return
    let scan = results.get(path)
    if (!scan) {
      try {
        scan = await scanner().scan(path, { deep: cfg().deep })
        results.set(scan)
      } catch (e: unknown) {
        vscode.window.showWarningMessage(`depwatch: ${e instanceof Error ? e.message : String(e)}`)
        return
      }
    }
    const uri = await vscode.window.showSaveDialog({
      filters: { SVG: ['svg'] },
      saveLabel: 'Export chart',
      defaultUri: defaultExportUri('depwatch-quadrant.svg'),
    })
    if (!uri) return
    const svg = quadrantSVG(scan.report.deps, {
      title: `${scan.report.file} — drift × viability (${scan.report.totalLibyears.toFixed(2)} libyears)`,
      thresholds: cfg().thresholds,
    })
    await vscode.workspace.fs.writeFile(uri, new TextEncoder().encode(svg))
  })

  // --- the baseline ---

  // Writing a baseline does not rescan: the reports are already in hand, and
  // accepting them is a filter over what they say.
  register('depwatch.writeBaseline', async () => {
    if (results.size === 0) await scanAll({ quiet: true })
    const uri = baseline.uri()
    if (!uri) {
      vscode.window.showWarningMessage('depwatch: a baseline needs an open workspace folder.')
      return
    }
    // The scans in `results` may already be filtered by an older baseline, so
    // the new one is written from a full rescan of what is cached — otherwise
    // accepting twice would quietly forget the first set.
    const full = await Promise.all(results.all().map((s) => scanner().scan(s.path, { deep: s.deep })))
    await baseline.write(uri, full)
    for (const scan of full) results.set(baseline.apply(scan))

    const pick = await vscode.window.showInformationMessage(
      `depwatch: accepted ${acceptableCount(full)} finding(s) into ${basename(uri.fsPath)}. Only what gets worse will show from now on.`,
      'Open baseline',
    )
    if (pick) await vscode.window.showTextDocument(uri)
  })

  register('depwatch.clearBaseline', async () => {
    const uri = baseline.uri()
    if (!uri) return
    try {
      await vscode.workspace.fs.delete(uri)
    } catch {
      vscode.window.showInformationMessage('depwatch: there was no baseline to clear.')
      return
    }
    baseline.forget()
    await scanAll({ quiet: true })
    vscode.window.showInformationMessage('depwatch: baseline cleared. Every finding is shown again.')
  })

  // --- the token ---

  register('depwatch.setGitHubToken', async () => {
    const token = await vscode.window.showInputBox({
      title: 'GitHub token for depwatch deep scans',
      prompt: 'Public-repo read access is enough. Stored in the OS keychain, never in settings.',
      password: true,
      ignoreFocusOut: true,
    })
    if (token === undefined) return
    if (token === '') {
      await deps.secrets.delete('depwatch.githubToken')
      delete process.env.GITHUB_TOKEN
      vscode.window.showInformationMessage('depwatch: GitHub token cleared.')
      return
    }
    await deps.secrets.store('depwatch.githubToken', token)
    process.env.GITHUB_TOKEN = token
    vscode.window.showInformationMessage('depwatch: GitHub token stored. Deep scans will use it.')
  })
}

/**
 * The trend, or the one failure the editor can actually fix. A missing binary
 * is offered as a command rather than only named — and it runs in a terminal
 * and not through exec, because `sudo` needs somewhere to ask for a password
 * and a package manager mid-install is something you want to watch.
 */
async function showHistory(rel: string, cwd: string, cfg: Config): Promise<void> {
  try {
    const points = await trend(rel, undefined, {
      cwd,
      maxPoints: cfg.trendMaxPoints,
      thresholds: cfg.thresholds,
    })
    showTrend(rel, points)
  } catch (e: unknown) {
    const message = `depwatch: ${e instanceof Error ? e.message : String(e)}`
    if (!isMissingGit(e)) {
      vscode.window.showWarningMessage(message)
      return
    }
    const pick = await vscode.window.showWarningMessage(message, `Run \`${INSTALL_GIT}\``)
    if (!pick) return
    const terminal = vscode.window.createTerminal('depwatch: install git')
    terminal.show()
    terminal.sendText(INSTALL_GIT)
  }
}

function defaultExportUri(name: string): vscode.Uri | undefined {
  const folder = vscode.workspace.workspaceFolders?.[0]
  return folder ? vscode.Uri.joinPath(folder.uri, name) : undefined
}

File: extensions/nvim/lua/depwatch/commands.lua
-- The user commands, and the pickers in front of them.
--
-- Nothing here keeps state: `ctx` is the session -- its configuration, what the
-- last scan found, and the two calls that reach the CLI -- handed over by
-- init.lua, which owns all of it. That keeps a command readable on its own and
-- keeps one copy of the state, where the scan policy can see it.
--
---@class DepwatchContext
---@field cfg table the resolved configuration
---@field grouping fun():string how the report is bucketed this session
---@field set_grouping fun(mode:string)
---@field root fun():string
---@field manifests fun():string[]
---@field is_manifest fun(path:string):boolean
---@field results fun():table[] every scan, ordered by label
---@field scan_of fun(path:string):table|nil the scan of exactly this file
---@field totals fun():table
---@field scan fun(path:string, opts:table)
---@field scan_all fun(opts:table)
---@field log_lines fun():string[]
---@field log fun(fmt:string, ...any) append a line to the plugin's own log

local cli = require('depwatch.cli')
local core = require('depwatch.core')
local notify = require('depwatch.notify')
local quickfix = require('depwatch.quickfix')
local ui = require('depwatch.ui')

local M = {}

--- What each axis answers, for the picker. depwatch has no rule engine to group
--- by -- the two axes are measured, not asserted -- so severity is the quadrant
--- and ecosystem is the third real axis.
local GROUP_BLURB = {
  file = 'what is wrong in each manifest',
  severity = 'what is worst, across every manifest',
  ecosystem = 'which registry the drift is coming from',
}
--- The manifest a command should act on: the current buffer when it is one,
--- otherwise the only one, otherwise ask.
local function pick_manifest(ctx, cb)
  local current = vim.fs.normalize(vim.api.nvim_buf_get_name(0))
  if current ~= '' and ctx.is_manifest(current) then
    return cb(current)
  end
  local scans = ctx.results()
  local paths = {}
  for _, scan in ipairs(scans) do
    paths[#paths + 1] = scan.path
  end
  if #paths == 0 then
    paths = ctx.manifests()
  end
  if #paths == 0 then
    notify('no dependency manifest found.', vim.log.levels.WARN)
    return
  end
  if #paths == 1 then
    return cb(paths[1])
  end
  vim.ui.select(paths, {
    prompt = 'Which manifest?',
    format_item = function(path)
      return vim.fs.relpath(ctx.root(), path) or path
    end,
  }, function(choice)
    if choice then
      cb(choice)
    end
  end)
end

local function ensure_scanned(ctx, cb)
  if #ctx.results() > 0 then
    return cb()
  end
  notify('scanning…')
  ctx.scan_all({ on_all_done = cb })
end

--- The report, in a float.
function M.report(ctx)
  ensure_scanned(ctx, function()
    local scans = ctx.results()
    if #scans == 0 then
      return notify('nothing scanned yet.', vim.log.levels.WARN)
    end
    ui.float(ui.report_lines(scans, ctx.grouping()), {
      title = string.format(' depwatch report — by %s ', ctx.grouping()),
      filetype = 'depwatch-report',
    })
  end)
end

--- How the report buckets its rows. Config sets the default and this changes it
--- for the session: like the filter, it is a way of looking at today's report
--- rather than a setting worth persisting.
function M.group_by(ctx, mode)
  if mode == nil or mode == '' then
    return vim.ui.select(core.GROUP_BY, {
      prompt = 'Group the report by',
      format_item = function(item)
        return string.format('%-10s %s', item, GROUP_BLURB[item])
      end,
    }, function(choice)
      if choice then
        M.group_by(ctx, choice)
      end
    end)
  end
  if not vim.tbl_contains(core.GROUP_BY, mode) then
    return notify(
      string.format('cannot group by "%s" — try %s.', mode, table.concat(core.GROUP_BY, ', ')),
      vim.log.levels.WARN
    )
  end
  ctx.set_grouping(mode)
  M.report(ctx)
end

--- The editor's copy of `depwatch check --ci`.
function M.gates(ctx)
  if ctx.cfg.gates.max_libyears == nil and ctx.cfg.gates.max_replace == nil then
    return notify(
      'no gates configured. Set gates.max_libyears or gates.max_replace to fail on a budget, '
        .. 'the same way `depwatch check --ci` does.',
      vim.log.levels.WARN
    )
  end
  ensure_scanned(ctx, function()
    local failures = {}
    for _, scan in ipairs(ctx.results()) do
      for _, failure in ipairs(core.gate_failures(scan.report, ctx.cfg.gates)) do
        failures[#failures + 1] = scan.label .. ': ' .. failure.message
      end
    end
    if #failures == 0 then
      return notify('gates pass.')
    end
    notify(string.format('%d gate(s) failing —\n%s', #failures, table.concat(failures, '\n')), vim.log.levels.WARN)
  end)
end

--- One row per sampled commit, and the drift between the ends. Two commits are
--- the fewest that have a "between", so a single point gets no delta line.
local function trend_lines(label, points)
  local lines = { '# ' .. label, '' }
  for _, point in ipairs(points) do
    lines[#lines + 1] = string.format(
      '%s  %s  %8.2f libyears  %4d deps  %d replace',
      point.date:sub(1, 10),
      point.commit,
      point.totalLibyears,
      point.deps,
      point.replace
    )
  end
  if #points > 1 then
    local delta = points[#points].totalLibyears - points[1].totalLibyears
    lines[#lines + 1] = ''
    lines[#lines + 1] = string.format(
      '%s%.2f libyears over %d sampled commits',
      delta >= 0 and '+' or '',
      delta,
      #points
    )
  end
  return lines
end

--- Drift over the manifest's git history.
function M.trend(ctx)
  pick_manifest(ctx, function(path)
    local label = vim.fs.relpath(ctx.root(), path) or path
    notify('reading history of ' .. label .. '…')
    cli.run(ctx, core.trend_argv(ctx.cfg, path), function(out)
      local points, err = cli.decode(out)
      if not points or type(points) ~= 'table' or #points == 0 then
        return notify(err or (out.stderr ~= '' and out.stderr) or 'no history for that manifest.', vim.log.levels.WARN)
      end
      ui.float(trend_lines(label, points), { title = ' depwatch trend ' })
    end)
  end)
end

--- Accept everything currently found. Writes the same file
--- `depwatch check --accepted` reads, so CI and the editor agree.
function M.write_baseline(ctx)
  pick_manifest(ctx, function(path)
    cli.run(ctx, core.write_baseline_argv(ctx.cfg, path), function(out)
      local message = (out.stderr or ''):gsub('%s+$', '')
      if out.code ~= 0 then
        return notify(message ~= '' and message or 'could not write the baseline.', vim.log.levels.ERROR)
      end
      notify(message ~= '' and message:gsub('^depwatch: ', '') or ('wrote ' .. ctx.cfg.baseline.path))
      ctx.scan(path, {})
    end)
  end)
end

function M.clear_baseline(ctx)
  local target = vim.fs.joinpath(ctx.root(), ctx.cfg.baseline.path)
  if vim.uv.fs_stat(target) == nil then
    return notify('there was no baseline to clear.')
  end
  local ok, err = vim.uv.fs_unlink(target)
  if not ok then
    return notify('could not remove the baseline: ' .. tostring(err), vim.log.levels.ERROR)
  end
  notify('baseline cleared. Every finding is shown again.')
  ctx.scan_all({})
end

-- --- the quickfix list -------------------------------------------------------

--- Every finding, in the quickfix list. See quickfix.lua for what a row says.
function M.list(ctx, opts)
  opts = opts or {}
  ensure_scanned(ctx, function()
    quickfix.list(ctx.results(), ctx.cfg.thresholds, opts.all)
  end)
end

--- Show only some quadrants.
function M.filter(ctx)
  quickfix.filter(ctx.results, ctx.cfg.thresholds, ctx.totals())
end

--- What depwatch knows about the dependency under the cursor.
function M.hover(ctx)
  local path = vim.fs.normalize(vim.api.nvim_buf_get_name(0))
  local scan = ctx.scan_of(path)
  if not scan then
    return notify('this file has not been scanned.', vim.log.levels.WARN)
  end
  local text = table.concat(vim.api.nvim_buf_get_lines(0, 0, -1, false), '\n')
  local names = {}
  for _, dep in ipairs(scan.report.deps or {}) do
    names[#names + 1] = dep.name
  end
  local spans = core.locate(text, path, names)
  local lnum = vim.api.nvim_win_get_cursor(0)[1] - 1
  for _, dep in ipairs(scan.report.deps or {}) do
    local span = spans[dep.name]
    if span and span.lnum == lnum then
      return ui.hover(core.hover_lines(dep, ctx.cfg.thresholds, scan.report.ecosystem))
    end
  end
  notify('no dependency on this line.')
end

function M.show_log(ctx)
  local log_lines = ctx.log_lines()
  ui.float(#log_lines > 0 and log_lines or { 'nothing logged yet' }, { title = ' depwatch log ', filetype = 'log' })
end

return M
